services/src/tasks/mod.rs

```diff
@@ -0,0 +1,141 @@
+
+    async fn status(&self, task_id: TaskId) -> Result<TaskStatus, TaskError>;
+
+    async fn list(&self, options: TaskListOptions) -> Result<Vec<TaskStatusWithId>, TaskError>;
+}
+
+identifier!(TaskId);
+
+/// A task that can run asynchronously and reports its status.
+#[async_trait::async_trait]
+pub trait Task<C: TaskContext>: Send + Sync {
+    async fn run(self: Box<Self>, ctx: C) -> Result<Box<dyn TaskStatusInfo>, Box<dyn ErrorSource>>;
+
+    fn boxed(self) -> Box<dyn Task<C>>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+
+/// A way to supply status updates from a `Task` to a `TaskDb`.
+#[async_trait::async_trait]
+pub trait TaskContext: Send + Sync {
+    /// Set the completion percentage (%) of the task.
+    /// This is a number between 0 and 100.
+    ///
+    /// Moreover, set a status message.
+    ///
+    async fn set_completion(&self, pct_complete: u8, status: Box<dyn TaskStatusInfo>);
+}
+
+/// One of the statuses a `Task` can be in.
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase", tag = "status")]
+pub enum TaskStatus {
+    // Pending, // TODO: at some point, don't just run things
+    Running(Arc<RunningTaskStatusInfo>),
+    Completed {
+        info: Arc<Box<dyn TaskStatusInfo>>,
+    },
+    #[serde(serialize_with = "serialize_failed_task_status")]
+    Failed {
+        error: Arc<Box<dyn ErrorSource>>,
+    },
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct TaskStatusWithId {
+    pub task_id: TaskId,
+    #[serde(flatten)]
+    pub status: TaskStatus,
+}
+
+impl TaskStatus {
+    pub fn completed(info: Arc<Box<dyn TaskStatusInfo>>) -> Self {
+        TaskStatus::Completed { info }
+    }
+
+    pub fn failed(error: Arc<Box<dyn ErrorSource>>) -> Self {
+        TaskStatus::Failed { error }
+    }
+
+    pub fn is_running(&self) -> bool {
+        matches!(self, TaskStatus::Running(_))
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct RunningTaskStatusInfo {
+    pct_complete: u8,
+    info: Box<dyn TaskStatusInfo>,
+}
+
+impl RunningTaskStatusInfo {
+    pub fn new(pct_complete: u8, info: Box<dyn TaskStatusInfo>) -> Arc<Self> {
+        Arc::new(RunningTaskStatusInfo { pct_complete, info })
+    }
+}
+
+#[allow(clippy::borrowed_box)]
+fn serialize_failed_task_status<S>(
+    error: &Box<dyn ErrorSource>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    serializer.serialize_str(error.to_string().as_str())
+}
+
+/// Trait for information about the status of a task.
+pub trait TaskStatusInfo: erased_serde::Serialize + Send + Sync + fmt::Debug {
+    fn boxed(self) -> Box<dyn TaskStatusInfo>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+
+erased_serde::serialize_trait_object!(TaskStatusInfo);
+
+impl TaskStatusInfo for () {}
+impl TaskStatusInfo for String {}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TaskListOptions {
```

Comment:
what about `order` to get stable outputs?

Reply:
I don't know good ordering options, here. For now, it is stable, it is in the order of insertion.

Reply:
since when is iterating a hashmap stably sorted?

Reply:
I guess you forgot the implementation.

There is the following struct:

```Rust
pub struct SimpleTaskManager {
    handles: Db<HashMap<TaskId, JoinHandle<()>>>,
    tasks_by_id: Db<HashMap<TaskId, Db<TaskStatus>>>,
    task_list: Db<VecDeque<TaskUpdateStatusWithTaskId>>,
}
```

I use `tasks_by_id` for a fast lookup and `task_list` for listing the stuff. And a `VecDeque` has an order.